File: web/wire.js
// relay wire codecs — the browser mirror of internal/protocol/wire.go.
// Message-kind numbers are part of the contract (see PROTOCOL.md) and must match Go.

export const Kind = {
  Hello: 0x01,
  Output: 0x02,
  Exit: 0x03,
  Control: 0x04,
  Input: 0x10,
  Resize: 0x11,
  CtrlReq: 0x12,
  CtrlRel: 0x13,
};

export const Control = { ReadOnly: 0, Granted: 1 };

export function decodeHello(b) {
  if (b.length < 12) throw new Error("short hello");
  const dv = new DataView(b.buffer, b.byteOffset, b.byteLength);
  return { baseline: Number(dv.getBigUint64(0)), cols: dv.getUint16(8), rows: dv.getUint16(10) };
}

export function encodeResize(cols, rows) {
  const b = new Uint8Array(4);
  const dv = new DataView(b.buffer);
  dv.setUint16(0, cols);
  dv.setUint16(2, rows);
  return b;
}

export function decodeExit(b) {
  if (b.length < 4) throw new Error("short exit");
  return new DataView(b.buffer, b.byteOffset, b.byteLength).getInt32(0);
}

export function decodeControl(b) {
  return b.length > 0 ? b[0] : 0;
}

// EMPTY is the shared empty payload for control-request/release frames.
export const EMPTY = new Uint8Array(0);

File: internal/protocol/wire.go
package protocol

import (
	"encoding/binary"
	"errors"
)

// Message kinds. Low byte space is runner→viewer, high is viewer→runner; see
// PROTOCOL.md. The numbers are part of the contract and must not change.
const (
	KindHello   byte = 0x01 // runner→viewer: baseline seq + initial size
	KindOutput  byte = 0x02 // runner→viewer: raw PTY output
	KindExit    byte = 0x03 // runner→viewer: command exit code
	KindControl byte = 0x04 // runner→viewer: control state (read-only / granted)

	KindInput   byte = 0x10 // viewer→runner: raw keystrokes
	KindResize  byte = 0x11 // viewer→runner: terminal size
	KindCtrlReq byte = 0x12 // viewer→runner: request control
	KindCtrlRel byte = 0x13 // viewer→runner: release control
)

// Control states carried by KindControl.
const (
	ControlReadOnly byte = 0
	ControlGranted  byte = 1
)

var errShortPayload = errors.New("protocol: payload too short")

// EncodeHello builds a HELLO payload: the input-seq baseline the viewer must start
// at, plus the current terminal size.
func EncodeHello(baseline uint64, cols, rows uint16) []byte {
	b := make([]byte, 12)
	binary.BigEndian.PutUint64(b[0:8], baseline)
	binary.BigEndian.PutUint16(b[8:10], cols)
	binary.BigEndian.PutUint16(b[10:12], rows)
	return b
}

// DecodeHello parses a HELLO payload.
func DecodeHello(b []byte) (baseline uint64, cols, rows uint16, err error) {
	if len(b) < 12 {
		return 0, 0, 0, errShortPayload
	}
	return binary.BigEndian.Uint64(b[0:8]),
		binary.BigEndian.Uint16(b[8:10]),
		binary.BigEndian.Uint16(b[10:12]), nil
}

// EncodeResize builds a RESIZE payload.
func EncodeResize(cols, rows uint16) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint16(b[0:2], cols)
	binary.BigEndian.PutUint16(b[2:4], rows)
	return b
}

// DecodeResize parses a RESIZE payload.
func DecodeResize(b []byte) (cols, rows uint16, err error) {
	if len(b) < 4 {
		return 0, 0, errShortPayload
	}
	return binary.BigEndian.Uint16(b[0:2]), binary.BigEndian.Uint16(b[2:4]), nil
}

// EncodeExit builds an EXIT payload.
func EncodeExit(code int32) []byte {
	b := make([]byte, 4)
	binary.BigEndian.PutUint32(b, uint32(code))
	return b
}

// DecodeExit parses an EXIT payload.
func DecodeExit(b []byte) (int32, error) {
	if len(b) < 4 {
		return 0, errShortPayload
	}
	return int32(binary.BigEndian.Uint32(b)), nil
}
